core/src/main/scala/org/bykn/bosatsu/TypedExpr.scala

```diff
@@ -127,14 +158,70 @@ object TypedExpr {
         }
     }
 
-  def coerceFn(arg: rankn.Type, result: rankn.Type.Rho, coarg: Coerce, cores: Coerce): Coerce =
+  /**
+   * Return the list of the free vars
+   */
+  def freeVars[A](ts: List[TypedExpr[A]]): List[String] = {
+
+    // usually we can recurse in a loop, but sometimes not
+    def cheat(ts: List[TypedExpr[A]], bound: Set[String], acc: List[String]): List[String] =
+      go(ts, bound, acc)
+
+    @annotation.tailrec
+    def go(ts: List[TypedExpr[A]], bound: Set[String], acc: List[String]): List[String] =
+      ts match {
+        case Nil => acc
+        case Generic(_, expr, _) :: tail =>
+          go(expr :: tail, bound, acc)
+        case Annotation(t, _, _) :: tail =>
+          go(t :: tail, bound, acc)
+        case Var(name, _, _) :: tail if bound(name) => go(tail, bound, acc)
+        case Var(name, _, _) :: tail => go(tail, bound, name :: acc)
+        case AnnotatedLambda(arg, _, res, _) :: tail =>
+          val acc1 = cheat(res :: Nil, bound + arg, acc)
+          go(tail, bound, acc1)
+        case App(fn, arg, _, _) :: tail =>
+          go(fn :: arg :: tail, bound, acc)
+        case Let(arg, argE, in, _) :: tail =>
+          val acc1 = cheat(in :: Nil, bound + arg, acc)
+          go(argE :: tail, bound, acc1)
+        case Literal(_, _, _) :: tail =>
+          go(tail, bound, acc)
+        case If(c, ift, iff, _) :: tail =>
+          go(c :: ift :: iff :: tail, bound, acc)
+        case Match(arg, branches, _) :: tail =>
+          go(arg :: branches.toList.map(_._2) ::: tail, bound, acc)
+      }
+
+    ts.foldLeft(List.empty[String]) { (acc, t) =>
+      go(t :: Nil, Set.empty, acc)
+    }
+    // reverse and distinct in one go
+    .foldLeft((Set.empty[String], List.empty[String])) {
+      case (res@(vset, vs), v) if vset(v) => res
+      case ((vset, vs), v) => (vset + v, v :: vs)
+    }
+    ._2
+  }
+
+  /**
+   * TODO this seems pretty expensive to blindly apply: we are deoptimizing
+   * the nodes pretty heavily
+   */
+  def coerceFn(arg: Type, result: Type.Rho, coarg: Coerce, cores: Coerce): Coerce =
     new FunctionK[TypedExpr, TypedExpr] { self =>
-      def apply[A](expr: TypedExpr[A]) =
-        AnnotatedLambda("x", arg, cores(App(expr, coarg(Var("x", arg, expr.tag)), result, expr.tag)), expr.tag)
+      def apply[A](expr: TypedExpr[A]) = {
+        /**
+         * We have to be careful not to collide with the free vars in expr
+         */
+        val free = freeVars(expr :: Nil).toSet
+        val name = Type.allBinders.iterator.map(_.name).filterNot(free).next
```

Comment:
this was the bug: we can't blindly assign variable names because we may be shadowing if we do so (obviously). I copied the paper here without much thought.

I think the TypedExpr work is still somewhat weak.